pulsar-client/src/main/java/org/apache/pulsar/client/impl/GeoReplicationProducerImpl.java

```diff
@@ -0,0 +1,248 @@
+            return;
+        }
+        synchronized (this) {
+            OpSendMsg op = pendingMessages.peek();
+            if (op == null) {
+                if (log.isDebugEnabled()) {
+                    log.debug("[{}] [{}] Got ack for timed out msg {}:{}", topic, producerName, seq, highSeq);
+                }
+                return;
+            }
+            // Replicator send markers also, use sequenceId to check the marker send-receipt.
+            if (isReplicationMarker(highSeq)) {
+                ackReceivedReplMarker(cnx, op, seq, highSeq, ledgerId, entryId);
+                return;
+            }
+            ackReceivedReplicatedMsg(cnx, op, seq, highSeq, ledgerId, entryId);
+        }
+    }
+    private void ackReceivedReplicatedMsg(ClientCnx cnx, OpSendMsg op, long sourceLId, long sourceEId,
+                                          long targetLId, long targetEid) {
+        // Parse source cluster's entry position.
+        Long pendingLId = null;
+        Long pendingEId = null;
+        List<KeyValue> kvPairList =  op.msg.getMessageBuilder().getPropertiesList();
+        for (KeyValue kvPair : kvPairList) {
+            if (kvPair.getKey().equals(MSG_PROP_REPL_SOURCE_POSITION)) {
+                if (!kvPair.getValue().contains(":")) {
+                    break;
+                }
+                String[] ledgerIdAndEntryId = kvPair.getValue().split(":");
+                if (ledgerIdAndEntryId.length != 2 || !StringUtils.isNumeric(ledgerIdAndEntryId[0])
+                        || !StringUtils.isNumeric(ledgerIdAndEntryId[1])) {
+                    break;
+                }
+                pendingLId = Long.valueOf(ledgerIdAndEntryId[0]);
+                pendingEId = Long.valueOf(ledgerIdAndEntryId[1]);
+                break;
+            }
+        }
+
+        // Case-1: repeatedly publish. Source message was exactly resend by the Replicator after a cursor rewind.
+        //   - The first time: Replicator --M1--> producer --> ...
+        //   - Cursor rewind.
+        //   - The second time: Replicator --M1--> producer --> ...
+        if (pendingLId != null && pendingEId != null
+                && (pendingLId < lastPersistedSourceLedgerId || (pendingLId.longValue() == lastPersistedSourceLedgerId
+                  && pendingEId.longValue() <= lastPersistedSourceEntryId))) {
+            if (log.isDebugEnabled()) {
+                log.debug("[{}] [{}] Received an msg send receipt[pending send is repeated due to repl cursor rewind]:"
+                                + " source entry {}:{}, pending send: {}:{}, latest persisted: {}:{}",
+                        topic, producerName, sourceLId, sourceEId, pendingLId, pendingEId,
+                        lastPersistedSourceLedgerId, lastPersistedSourceEntryId);
+            }
+            removeAndApplyCallback(op, sourceLId, sourceEId, targetLId, targetEid, false);
+            ackReceived(cnx, sourceLId, sourceEId, targetLId, targetEid);
```

Comment:
Why does it need to call the method ackReceived?

Reply:
This logic branch is used for the following case:
- Replicator sends msg (origin-position: `3:1`)
  - Pending msg queue of the internal producer of the replicator is `[3:1]` now. 
- Replicator sends msg (origin-position: `3:1`), it repeats send the same msg
  -  Pending msg queue of the internal producer of replicator is `[3:1], [3:1]` now.
- This logic branch will remove the first record in Pending msg queue of the internal producer

Reply:
Got it, thanks.